### ISSUE TITLE
Unble to install any ruby version through rvm on Monterey M1 Mac

### ISSUE THREAD

**Issue by pujith22** [rvm/rvm#5153]
## Description

Unable to install any ruby version through rvm on Monterey on M1 Mac
getting the error "Error running '__rvm_make -j8', every time,

## Steps to reproduce

 1. rvm install 2.5.0 or rvm install 2.3.6
 

## Expected behavior

ruby to be installed successfully

## Actual behavior
    Build failing (unable to complete the installation) always throwing the following error "Error running '__rvm_make -j8'

* [command](https://gist.github.com/...)
[make.log](https://github.com/rvm/rvm/files/7525208/make.log)
[configure.log](https://github.com/rvm/rvm/files/7525213/configure.log)

**Debug mode**:
* [command](https://gist.github.com/...)
* [config.log](https://gist.github.com/...) - only for issues with `install`
* [make.log](https://gist.github.com/...) - only for issues with `install`
*

## Environment info

system:
xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance

  system:
    uname:        "Darwin PUJITHs-MacBook-Air.local 21.1.0 Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:24 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T8101 arm64"
    name:         "OSX"
    version:      "12.0"
    architecture: "arm64"
    bash:         "/opt/homebrew/bin/bash => GNU bash, version 5.1.8(1)-release (aarch64-apple-darwin20.4.0)"
    zsh:          "/bin/zsh => zsh 5.8 (x86_64-apple-darwin21.0)"
    remote_path:  "osx/12.0/arm64"
    xcode:        ""

  rvm:
    version:      "1.29.12-next (master)"
    updated:      "17 hours 55 minutes 52 seconds ago"
    path:         "/Users/pujith22/.rvm"
    autolibs:     "[4] Allow RVM to use package manager if found, install missing dependencies, install package manager (only OS X)."

  homes:
    gem:          "not set"
    ruby:         "not set"

  binaries:
    ruby:         "/usr/bin/ruby"
    irb:          "/usr/bin/irb"
    gem:          "/usr/bin/gem"
    rake:         "/usr/bin/rak

  environment:
    PATH:         "/Users/pujith22/.asdf/shims:/opt/homebrew/Cellar/asdf/0.8.1_1/libexec/bin:/opt/homebrew/opt/ruby@2.5/bin:/Users/pujith22/.rbenv/bin:/Users/pujith22/.nvm/versions/node/v16.4.2/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/pujith22/.rvm/bin"
    GEM_HOME:     ""
    GEM_PATH:     ""
    MY_RUBY_HOME: ""
    IRBRC:        ""
    RUBYOPT:      ""
    gemset:       ""

**Comment by mayur-kambariya** [comment]
1) Right clic Terminal from the Application/Utilities folder, Get Info, tick the "Open using Rosetta" box.
2) Uninstall Homebrew
  ```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall.sh)"
  rm -rf /opt/homebrew/*
  sudo rm -rf /opt/homebrew
```
3) Reinstall Homebrew
  `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
4) Restart terminal
5) Check Homebrew is working fine: `brew doctor`
6) Reinstall openssl: `brew install openssl`
7) Install Ruby: `rvm install 2.7` Or any version

**Comment by pujith22** [comment]
It worked man, thanks for the heads up. appreciate it

**Comment by mlch911** [comment]
> 1. Right clic Terminal from the Application/Utilities folder, Get Info, tick the "Open using Rosetta" box.
> 2. Uninstall Homebrew
> 
> ```
> /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall.sh)"
> rm -rf /opt/homebrew/*
> sudo rm -rf /opt/homebrew
> ```
> 
> 3. Reinstall Homebrew
>    `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
> 4. Restart terminal
> 5. Check Homebrew is working fine: `brew doctor`
> 6. Reinstall openssl: `brew install openssl`
> 7. Install Ruby: `rvm install 2.7` Or any version

This is not a good solution. You just install the x86 version using Rosetta. Any idea how to install it in M1 natively?

**Comment by ZASMan** [comment]
> 

Did you find a solution for this?

**Comment by ATL-Flaneur** [comment]
I gave up on RVM for the time being and installed Ruby through MacPorts. The build recipe sorts out OpenSSL. Same with Homebrew if that's more your style.

**Comment by pujith22** [comment]
Yeah, so far I haven't found a way to install RVM for ARM architecture..
Still using x86 version only and that works for me as of now.. Will update
on it if I find any alternative.

**Comment by mlch911** [comment]
I gave up on RVM and use Rbenv instead. Rbenv works fine.

**Comment by mfaughn** [comment]
Had same problem with brand new M1 Max MBP w/ MacOS 12.3.  Tried rbenv and also got failure.  Had success with `RUBY_CFLAGS=-DUSE_FFI_CLOSURE_ALLOC arch -arm64 rbenv install 2.7.1`.  Maybe that helps someone figure out how to get RVM to work?

**Comment by matthewhively** [comment]
Why is this closed without an actual solution?

**Comment by matthewhively** [comment]
Relevant issues we found:
https://github.com/rvm/rvm/issues/4968
https://github.com/libffi/libffi/issues/571

**Comment by mikel** [comment]
As a note, I had this issue, was caused by doing an upgrade from an intel mac to an M1 mac. Solved by uninstalling RVM and Homebrew completely, then reinstalling fresh.  Took a little while but the rubies are now isntalling cleanly without lots of errors.

Key one to install was `brew install ruby-build` then checking any errors and installing other libs like `brew install libyaml` etc.

**Comment by imrichardcole** [comment]
Same issue for me, removing homebrew and rvm and starting from scratch didn't help sadly

**Comment by imrichardcole** [comment]
I've stepped through all the versions of ruby given by `rvm list known` (not going back further than 2.0.0) and produced the follow table of success/fail:

2.0.0:	fail
2.1.10:	fail
2.2.10:	fail
2.3.8:	fail
2.4.10:	fail
2.5.8:	fail
2.6.6:	fail
2.7.2:	success
3.0.0:	success

rvm version as follows:

rvm 1.29.12 (latest) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io]

**Comment by pkuczynski** [comment]
As I do not have mac M1, its impossible to test easily. We would however welcome any PRs from the community to make it work. If you see there are dependencies missing, you can add them to https://github.com/rvm/rvm/blob/master/scripts/functions/requirements/osx_brew - in there you can also make more complex logic depending on the architecture or os version. Examples can be found in other files in this folder, eg https://github.com/rvm/rvm/blob/master/scripts/functions/requirements/ubuntu#L47

**Comment by Meekohi** [comment]
Worked for me (from https://github.com/rvm/rvm/issues/4968):

```
brew install libffi
export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig"
LDFLAGS="-L/opt/homebrew/opt/libffi/lib" CPPFLAGS="-I/opt/homebrew/opt/libffi/include" rvm install
```

**Comment by pkuczynski** [comment]
@Meekohi would you like to come up with PR maybe?

**Comment by Meekohi** [comment]
Not sure where to start alas -- it seems to me the core problem is a dependency on an x86-only version of `libffi` but I don't see any version pinned by `rvm` so I'm not really certain how to enforce that `libffi` be up-to-date or reinstalled.

**Comment by pkuczynski** [comment]
Everything is here https://github.com/rvm/rvm/blob/master/scripts/functions/requirements/osx_brew. You can add `libffi` to `requirements_osx_brew_libs_default`. I am not sure if defining these extra flags would be necessary. You can start without them and see...

**Comment by payswithfifties** [comment]
I'm currently trying to see how to compile ruby from source (without RVM) once I have a version compiling, it should reveal where the problem lies (and hopefully the solution!)

**Comment by Meekohi** [comment]
In my so-far failed attempts to propose a PR I did learn a few things I'll document here. 

The core problem is that the version of libffi you currently have installed will trigger this error:
```
../../symbol.h:22:71: note: expanded from macro 'rb_id2sym':264:14: error: implicit declaration of function 'ffi_prep_closure' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
```

## Work Around 1
Installing a newer version via homebrew isn't quite sufficient since the compiler doesn't know to use that new location by default, so you need the extra flags:
```
brew install libffi
export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig"
LDFLAGS="-L/opt/homebrew/opt/libffi/lib" CPPFLAGS="-I/opt/homebrew/opt/libffi/include" rvm install 2.5.1
```

## Work Around 2
An alternative work-around is to do
```
CPPFLAGS=-DUSE_FFI_CLOSURE_ALLOC rvm install 2.5.1 
```
which tells the compiler to go along with the invalid syntax, and continue using the old version of libffi.

## Work Around 3
Apparently this all stems from fiddle, which I imagine the vast majority of people don't use, so another option is.

[Fiddle](https://github.com/ruby/fiddle) is an extension to translate a foreign function interface (FFI) with ruby.
```
rvm install 2.5.1 --with-out-ext=fiddle
```

**Comment by julestruong** [comment]
None of the above solutions worked for me :( 
(rvm install 2.6.5 tho)

**Comment by danielricecodes** [comment]
> Worked for me (from #4968):
> 
> ```
> brew install libffi
> export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig"
> LDFLAGS="-L/opt/homebrew/opt/libffi/lib" CPPFLAGS="-I/opt/homebrew/opt/libffi/include" rvm install 2.5.1
> ```

Worked!  Yay!!!!  🎉

**Comment by simonknoll** [comment]
For me, uninstalling brew and rvm and reinstalling it did the trick:
https://stackoverflow.com/a/74443264/328602

**Comment by muratatak77** [comment]
> > Worked for me (from #4968):
> > ```
> > brew install libffi
> > export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig"
> > LDFLAGS="-L/opt/homebrew/opt/libffi/lib" CPPFLAGS="-I/opt/homebrew/opt/libffi/include" rvm install 2.5.1
> > ```
> 
> Worked! Yay!!!! 🎉

Worked for me also. 
Using MacOS Ventura 13.2

**Comment by SanjanaTailor** [comment]
MacOS Monterey version 12.5 
command > `rvm install 2.7.1`
```
Searching for binary rubies, this might take some time.
No binary rubies available for: osx/12.5/arm64/ruby-2.5.1.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for osx.

Updating certificates bundle '/opt/homebrew/etc/openssl@1.1/cert.pem'
Requirements installation successful.
Installing Ruby from source to: /Users/sanjanaben.tailor/.rvm/rubies/ruby-2.5.1, this may take a while depending on your cpu(s)...
ruby-2.5.1 - #downloading ruby-2.5.1, this may take a while depending on your connection...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13.3M  100 13.3M    0     0  1150k      0  0:00:11  0:00:11 --:--:-- 1156k
ruby-2.5.1 - #extracting ruby-2.5.1 to /Users/sanjanaben.tailor/.rvm/src/ruby-2.5.1 - please wait
ruby-2.5.1 - #applying patch /Users/sanjanaben.tailor/.rvm/patches/ruby/2.5.1/libressl_2_7.patch - please wait
ruby-2.5.1 - #configuring - please wait
ruby-2.5.1 - #post-configuration - please wait
ruby-2.5.1 - #compiling - please wait
Error running '__rvm_make -j10',
please read /Users/sanjanaben.tailor/.rvm/log/1675781537_ruby-2.5.1/make.log

There has been an error while running make. Halting the installation.
```
```
ruby --version 
ruby 2.6.8p205 (2021-07-07 revision 67951) [universal.arm64e-darwin21]
```

**Comment by SanjanaTailor** [comment]
I resolved this issue; 
cd .rvm/src/ruby-2.7.5 and 
run make install

**Comment by amritanupam** [comment]
I had this issue on mac M1 Pro Ventura. Steps to resolve:
I used normal installation of rvm according to rvm.io
brew install openssl@1.1
brew uninstall --ignore-dependencies openssl@3
rvm install 2.5.1
brew install --ignore-dependencies openssl@3

**Comment by philipborbon** [comment]
For anyone that stumbled on this from google search:

1. Confirm `OpenSSL 1.1.x` is enabled
```ssh
openssl version
```
You should have `OpenSSL 1.1.x`, if it showing other version, install `OpenSSL 1.1.x` first.

2. Install openssl version `OpenSSL 1.1.x`
```ssh
brew install openssl@1.1
```

3. Unlink current openssl version
```ssh
brew unlink openssl
```

4. Link `OpenSSL 1.1.x`
```
brew link openssl@1.1
```

5. Confirm you have linked `openssl@1.1`
```
openssl version
```
 
 6. Run your usual rvm install command
 ```
 rvm install ruby
 ```

**Comment by AzhirAhmadi** [comment]
> For anyone that stumbled on this from google search:
> 
> 1. Confirm `OpenSSL 1.1.x` is enabled
> 
> ```
> openssl version
> ```
> 
> You should have `OpenSSL 1.1.x`, if it showing other version, install `OpenSSL 1.1.x` first.
> 
> 2. Install openssl version `OpenSSL 1.1.x`
> 
> ```
> brew install openssl@1.1
> ```
> 
> 3. Unlink current openssl version
> 
> ```
> brew unlink openssl
> ```
> 
> 4. Link `OpenSSL 1.1.x`
> 
> ```
> brew link openssl@1.1
> ```
> 
> 5. Confirm you have linked `openssl@1.1`
> 
> ```
> openssl version
> ```
> 
> 6. Run your usual rvm install command
> 
> ```
> rvm install ruby
> ```

this one worked for 2.7.4

**Comment by tacianoscur** [comment]
> For anyone that stumbled on this from google search:
> 
> 1. Confirm `OpenSSL 1.1.x` is enabled
> 
> ```
> openssl version
> ```
> 
> You should have `OpenSSL 1.1.x`, if it showing other version, install `OpenSSL 1.1.x` first.
> 
> 2. Install openssl version `OpenSSL 1.1.x`
> 
> ```
> brew install openssl@1.1
> ```
> 
> 3. Unlink current openssl version
> 
> ```
> brew unlink openssl
> ```
> 
> 4. Link `OpenSSL 1.1.x`
> 
> ```
> brew link openssl@1.1
> ```
> 
> 5. Confirm you have linked `openssl@1.1`
> 
> ```
> openssl version
> ```
> 
> 6. Run your usual rvm install command
> 
> ```
> rvm install ruby
> ```

worked! thank you

**Comment by souzanato** [comment]
None of this worked for me until I downgraded xcode to version 13.

**Comment by nitishgupta-fareye** [comment]
> In my so-far failed attempts to propose a PR I did learn a few things I'll document here.
> 
> The core problem is that the version of libffi you currently have installed will trigger this error:
> 
> ```
> ../../symbol.h:22:71: note: expanded from macro 'rb_id2sym':264:14: error: implicit declaration 
> of function 'ffi_prep_closure' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
> ```
> 
> ## Work Around 1
> Installing a newer version via homebrew isn't quite sufficient since the compiler doesn't know to use that new location by default, so you need the extra flags:
> 
> ```
> brew install libffi
> export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig"
> LDFLAGS="-L/opt/homebrew/opt/libffi/lib" CPPFLAGS="-I/opt/homebrew/opt/libffi/include" rvm install 2.5.1
> ```
> 
> ## Work Around 2
> An alternative work-around is to do
> 
> ```
> CPPFLAGS=-DUSE_FFI_CLOSURE_ALLOC rvm install 2.5.1 
> ```
> 
> which tells the compiler to go along with the invalid syntax, and continue using the old version of libffi.
> 
> ## Work Around 3
> Apparently this all stems from fiddle, which I imagine the vast majority of people don't use, so another option is.
> 
> [Fiddle](https://github.com/ruby/fiddle) is an extension to translate a foreign function interface (FFI) with ruby.
> 
> ```
> rvm install 2.5.1 --with-out-ext=fiddle
> ```

Only workaround 3 worked for ruby 2.6.5 me. However, while running the rails console, it is looking for fiddle dependencies in the /.rvm/ directory that are not present.

**Comment by davidlj95** [comment]
FWIW I have a StackOverflow question about [installing any Ruby version in macOS with `rvm`][soq].

[soq]: https://stackoverflow.com/q/76922732/3263250

TL;DR: for recent versions [export `CONFIGURE_ARGS` env var as instructed per Ruby build guide](https://github.com/ruby/ruby/blob/30a5b94517699589f6943163cd6b92f2f6c0023f/doc/contributing/building_ruby.md#:%7E:text=typically%20using%20Homebrew%20on%20macOS). For older versions, [`ruby-build`](https://github.com/rbenv/ruby-build) can help dealing with Ruby's requirements. More details in the [StackOverflow question's answers][soq]
